GraphEdit/GraphNode refactor [V2]

Geometror opened the pull request:
This PR implements godotengine/godot-proposals#5271 for the most part.
Supersedes #61414 and #61783.

![image](https://user-images.githubusercontent.com/50084500/194773670-1042c46d-d33c-40f8-a45d-8cdbe2cc1fe9.png)

# Detailed changes
## Restructuring/Code style
- Split GraphNode into BaseGraphNode, GraphNode and GraphFrame (open for naming suggestions) due to:
	-   Maintainability: The GraphNode class got really large and complex, so splitting it up improves code navigation and readability
	-   Flexibility: Allows adding more features to the frame graph node or extend BaseGraphNode for a completely custom node without most of the predefined connection functionality (reference image backdrops, etc.)
	-   Performance/implementation complexity: For example, a comment node does not need to keep track of any connections/slots/ports
- Rename lots of short variable names to be more descriptive (e.g. icofs -> icon_offset, c -> child)
- Restructure the header and source files to follow a consistent scheme (i. a. order of member variable and method declarations, which were sometimes mixed)
- Fix comment style in various places
- GraphNode: Rename get_connection_input/output_* to get_port_input/output_*
- GraphEdit: Rename get_zoom_hbox to get_menu_hbox
- Add closable member variable to VisualShaderNode, before that show_close was used to save whether a GraphNode can be deleted (which is a bit hacky to begin with and show_close is gone)

## Functional changes
- GraphFrame
	- Tint color
		- Added context menu options to VisualShader editor 
	- GraphFrame can be grabbed and on all sides, not just on the titlebar
		- drag_margin property
	- title_centered property (for convenience, can be achieved via get_titlebar_hbox().get_child(0).horizontal_alignment = HORIZONTAL_ALIGNMENT_CENTER)
- The order of frame nodes is now calculated similar to a 2D BVH: when a comment node encloses other comment nodes, the index of the enclosing node is kept below (subject to change, see "future work" below)
    -   The order is updated when resizing a comment node
- The connections layer is now kept between the comment nodes and the normal graph
- Removed the close button (and its theming) from GraphNode by default (can be added manually if needed or by default with the new titlebar if we really want to keep it)
- Complete control over the titlebar of GraphFrame/GraphNode, e.g. custom controls (like a close or settings button, some status label) can be added to GraphFrame's/GraphNode's title bar via get_titlebar_hbox().add_child(...)
	- by default the titlebar hbox now holds a Label control

## Theming
- StyleBox for body (slot area) and title bar (both with selected variants)
- Use "GraphNodeTitleLabel" and "GraphFrameTitleLabel" theme variations for title labels, remove the title offset theme constants
- Adjustments to the editor theme and default project theme
![image](https://user-images.githubusercontent.com/50084500/194775274-1689a5be-9f60-4d4a-abda-cd6253c5e4d8.png)

## Fixes
- Fix custom port icon not being centered when it's bigger than the default icon
- VisualShader: Fix error label increasing the minimum size in an ugly way since it autowrap isn't enabled
- Fix and improve documentation with regard to port/connection/slot terminology

# Future work
In my view, there are still some things left to do, but I wanted to get this out now to receive some feedback.
These are some of the things that I will work on next or in the future, that might require some discussion (and more time). These may go in their own PRs or if desired/necessary, in this one.
- Change the GraphFrame behavior, as it currently has some problems and isn't really nice to work with in some cases (also requires complex update logic under the hood)
	- change it to work similar to Blender, where you can drop nodes in the frame (only then they are linked with the frame and move with it/extend its size, not just when you move the frame below them). Maybe actually utilize the scene tree for proper parent-child-relationship
	- add an option to shrink the GraphFrame based on its enclosing nodes
- Probably introduced with the change above, a layer index property for BaseGraphNode to improve/flesh out the concept of layers
- Evaluate the changes in this PR, gather feedback to improve upon
- Move the arranging logic out of GraphEdit in its own class (already over 500 LOC and it may get more complex)
- (?) maybe move the whole graph stuff in a module
- Provide the ability to minimize nodes
- Find ways to reduce some code duplicates (mostly in GraphEdit)
- Reevaluate the slot Stylebox

VexfulNiko commented:
Why did the Linux build fail?

fire commented:
I support this, and the work was started before the feature freeze.

QueenOfSquiggles commented:
I'm super excited about these features. It'll make my development of Choreographer 1000% easier. Great job @Geometror ! Hoping these can get accepted & merged

VexfulNiko commented:
I support this as well.

paddy-exe commented:
About the GraphFrame behaviour: IMO Blender has this done really nicely where it automatically detects when you put in the nodes inside the frame and resized itself to fit in the node after the drag finishes. The other points like e.g. the minimizing nodes feature would be a nice to have for larger graphs or subgraphs but not sure how much work this would be to implement. Would these changes be compatibility breaking or could they still be implemented when your rework PR is merged?
Also, again thank you for your work on this! It's really great how the changes are making the graph node system even more useful!

Geometror committed:
Splitup GraphNode into several classes

Geometror committed:
Theming and titlebar flexibility improvements

Geometror committed:
GraphEdit: Separate show grid from grid snap, small theme adjustments

joaopedrosgs commented:
Any news on this?

fire commented:
I would like this but Godot Engine 4.0 tasks have taken priority.

YuriSizov commented:
This PR, while it has some breaking changes, will be a part of 4.1. We'll see if we can smooth out the experience for existing users, though this node is unlikely to be used in games. It's more of a tool for editor plugins and standalone tools. This is to say that in the worst case scenario, developers should still be capable of updating their tools to the new setup.

Zireael07 commented:
I will chime in (tried using existing GraphEdit/Node for a tool) - this refactor is really needed

RabbitB commented:
> I will chime in (tried using existing GraphEdit/Node for a tool) - this refactor is really needed

I'm a proponent of making your own tools when it makes sense (to boost dev ease & speed). Off and on I've been working on a graphing calculator for expressions. I created it so I could quickly experiment and graph their output. I like to think of it as a visual sandbox to help with shader and other math-heavy development. It has been useful enough that I'd release it as a free utility, but while GraphEdit is useful for homebrew tools, it feels underdeveloped for something I would be comfortable putting my name on.

ajreckof commented on scene/gui/base_graph_node.cpp:
```diff
@@ -0,0 +1,245 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "graph_node.h"
```
```suggestion
#include "base_graph_node.h"
```
While incorporating with personal modification found this typo

ajreckof commented on scene/gui/graph_frame.cpp:
```diff
@@ -0,0 +1,270 @@
+	ClassDB::bind_method(D_METHOD("get_tint_color"), &GraphFrame::get_tint_color);
+
+	ADD_PROPERTY(PropertyInfo(Variant::STRING, "title"), "set_title", "get_title");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "title_centered"), "set_title_centered", "is_title_centered");
```
As `title` is shared between the two class this could be implemented in the `BaseGraphNode`as it would also benefit `GraphNode` (Wanted this especially while working on my plugin before discovering it was implemented in `GrephFrame`)

Geometror replied:
I didn't implement it in BaseGraphNode since that class should only provide the basic functionality for a node to be used as a GraphEdit child node (dragging/selecting/etc.). A title is needed for GraphNode and GraphFrame, but not necessarily for something like a image backdrop node or a reroute node. I'm still not sure what to do with the title_centered property because its just a convinience property, meaning that you can currently center the title of GraphFrame and GraphNode (and do many more things) with:
(in `_ready()` method)
```
var titlebar_hbox: Container = get_titlebar_hbox()
var label: Label = titlebar_hbox.get_child(0)
label.horizontal_alignment = HORIZONTAL_ALIGNMENT_CENTER
```
Since the method above is much more flexible I am thinking about removing tilte_centered.

ajreckof reviewed (commented):
While working on a plugin that uses GraphNode, I decided to try your PR. After rebasing I used it a bit. Here are a few things I thought while using it.

Geometror commented on scene/gui/graph_node.cpp:
```diff
@@ -1069,104 +844,59 @@ Vector<int> GraphNode::get_allowed_size_flags_vertical() const {
-	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "selectable"), "set_selectable", "is_selectable");
-	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "selected"), "set_selected", "is_selected");
-	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "comment"), "set_comment", "is_comment");
-	ADD_PROPERTY(PropertyInfo(Variant::INT, "overlay", PROPERTY_HINT_ENUM, "Disabled,Breakpoint,Position"), "set_overlay", "get_overlay");
-
-	ADD_GROUP("BiDi", "");
-	ADD_PROPERTY(PropertyInfo(Variant::INT, "text_direction", PROPERTY_HINT_ENUM, "Auto,Left-to-Right,Right-to-Left,Inherited"), "set_text_direction", "get_text_direction");
-	ADD_PROPERTY(PropertyInfo(Variant::STRING, "language", PROPERTY_HINT_LOCALE_ID, ""), "set_language", "get_language");
-	ADD_GROUP("", "");
```
overlay got removed as it was only used in VisualScript and hence very application specific. The other properties can be accessed via the titlebar label (see my other comment below on how to access titlebar nodes).

Geometror committed:
Extract GraphEdit node arrangement

Geometror commented:
Rebase and small update (lets call it 2.1):
- Separate snapping and grid (https://github.com/godotengine/godot-proposals/discussions/5250)
	- separate properties and toggle buttons (use_snap and show_grid):
![Pasted image 20221028165043](https://user-images.githubusercontent.com/50084500/230109472-d676906d-fe54-464c-a491-cf46afc74654.png)

- get_port_input/output_position returns no longer the position premultiplied with the scale
- `GraphEdit`'s node arrangement logic is now extracted into its own class `GraphEditArranger`; its coupling was loose enough so that only trivial adjustments were needed for it to use `GraphEdit`'s API
- Several fixes and further cleanup
- Rename `BaseGraphNode` to `GraphControl` and add an icon

Unfortunately, it's still not in a mergeable state since the sorting of GraphFrames is still broken and probably isn't fixable with the current/old behavior (but I think its ready for an initial review). I will experiment in the following days how a behavior like in Blender (where this problem is solved very elegantly) could be implemented in Godot.

berkeleynerd commented:
I'm working on a version of Trizbort IF mapper re-implemented in Godot and GraphNodes would serve my purpose well if ports could be enabled on the top and bottom of the node as well. (e.g., adding "Bottom Enabled" and "Top Enabled" in addition to the existing Left / Right)

I have no idea how much this would complicate auto-layout etc. but if I could implement this by creating a custom node extending the proposed BaseGraphNode that would certainly be useful as well.

paddy-exe commented:
When trying out this PR (again great work btw :D) I noticed that the Comment node is now called Frame. This change should also be applied to the Submenu `Set Comment Title`, `Set Comment Description`:
![image](https://user-images.githubusercontent.com/38077837/230746849-9055d577-413d-43a2-b063-fbb569fdaee6.png)
![image](https://user-images.githubusercontent.com/38077837/230746851-c55cefb7-ae60-4dd6-8a5a-667bb3a72b78.png)

YuriSizov reviewed (changes requested):
I'm starting to look at the code, but have a few general notes.

- GraphFrame is barely usable, and you also mention issues with it in your latest comments. My suggestion is to just drop it for now, remove that feature completely for the initial implementation. Its UX is not there yet, and it distracts you with layering issues. So let's get back to it in a follow-up.
  - Some fallback code needs to be added for visual shaders though, so that comments information is preserved, even if you can't see or edit it.

- I'm concerned about the naming scheme. IMO GraphControl doesn't make sense as the base type for GraphNode, since in Godot's terms Node is more basic than Control. So, I think it would be confusing with the concepts reversed. Perhaps GraphObject or GraphItem would be better as the base type?

- I'd urge you to not add more features to this PR. As I mentioned in the proposal before, this should be really done in steps. Some things are easy to add alongside the general refactoring, but the more changes this includes, the harder it is to get to a mergeable state and review. I'd like to have the initial refactoring merged within the next two weeks, so we can have it in 4.1.

There are also a couple of review comments. I'll look into the implementation soon.

YuriSizov commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -413,32 +418,49 @@ void VisualShaderGraphPlugin::add_node(VisualShader::Type p_type, int p_id, bool
 	Ref<Theme> vstheme;
 	vstheme.instantiate();
 	Ref<Font> label_font = EditorNode::get_singleton()->get_editor_theme()->get_font("main_msdf", "EditorFonts");
+	Ref<Font> label_bold_font = EditorNode::get_singleton()->get_editor_theme()->get_font("main_bold_msdf", "EditorFonts");
 	vstheme->set_font("font", "Label", label_font);
+	vstheme->set_font("font", "GraphNodeTitleLabel", label_bold_font);
 	vstheme->set_font("font", "LineEdit", label_font);
 	vstheme->set_font("font", "Button", label_font);
```
Not important right now, but worth extracting this into `editor_themes.cpp`. There isn't a good framework right now for creating additional themes in there, so that is going to require a bit of work.

YuriSizov commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -413,32 +418,49 @@ void VisualShaderGraphPlugin::add_node(VisualShader::Type p_type, int p_id, bool
-	GraphNode *node = memnew(GraphNode);
+	GraphControl *node;
+	if (is_frame) {
+		GraphFrame *frame = memnew(GraphFrame);
+		frame->set_title(vsnode->get_caption());
+		node = frame;
```
So to add to my suggestion before, I think a custom graph element can be put here in to temporarily store the data from existing visual shaders without rendering anything.

YuriSizov commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -1925,16 +1949,17 @@ void VisualShaderEditor::_draw_color_over_button(Object *obj, Color p_color) {
-void VisualShaderEditor::_update_created_node(GraphNode *node) {
-	const Ref<StyleBoxFlat> sb = node->get_theme_stylebox(SNAME("frame"), SNAME("GraphNode"));
-	Color c = sb->get_border_color();
+//TODO @Geometror: Solve this
```
Solve what? :)

Geometror replied:
Mmh, I solved it by removing the comment :)

I think this was a relict of an early version of this rework.

YuriSizov commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -3767,13 +3851,16 @@ void VisualShaderEditor::_convert_constants_to_parameters(bool p_vice_versa) {
-	List<int> to_erase;
-	to_erase.push_back(p_node);
+	Ref<VisualShaderNode> node = visual_shader->get_node((VisualShader::Type)p_type, p_node);
+	if (node->is_closable()) {
```
We generally prefer early returns to nesting.

YuriSizov reviewed (commented):
A bunch of comments on the API surface and on the changes in editor plugins.

YuriSizov commented on scene/gui/graph_control.cpp:
```diff
@@ -0,0 +1,245 @@
+
+#include "graph_edit.h"
+
+void GraphControl::_close_requested() {
+	// Send focus to parent.
+	get_parent_control()->grab_focus();
+	emit_signal(SNAME("close_request"));
+}
```
As mentioned before, this doesn't seem to be called anywhere.

YuriSizov commented on scene/gui/graph_edit.cpp:
```diff
@@ -270,78 +273,78 @@ void GraphEdit::_update_scroll_offset() {
 	for (int i = 0; i < get_child_count(); i++) {
-		GraphNode *gn = Object::cast_to<GraphNode>(get_child(i));
-		if (!gn) {
+		GraphControl *graph_node = Object::cast_to<GraphControl>(get_child(i));
```
The variable name is confusing as it refers to a different type.

YuriSizov commented on scene/gui/graph_edit.cpp:
```diff
@@ -270,78 +273,78 @@ void GraphEdit::_update_scroll_offset() {
 
-	if (!setting_scroll_ofs) { //in godot, signals on change value are avoided as a convention
-		emit_signal(SNAME("scroll_offset_changed"), get_scroll_ofs());
+	// In godot, signals on change value are avoided as a convention.
```
```suggestion
	// In Godot, signals on value change are avoided by convention.
```
🙃

YuriSizov commented on scene/gui/graph_edit.cpp:
```diff
@@ -270,78 +273,78 @@ void GraphEdit::_update_scroll_offset() {
 	for (int i = 0; i < get_child_count(); i++) {
-		GraphNode *gn = Object::cast_to<GraphNode>(get_child(i));
-		if (!gn) {
+		GraphControl *graph_node = Object::cast_to<GraphControl>(get_child(i));
```
Same issue with the variable name. Probably a bunch of others around all the touched files.

YuriSizov commented on scene/gui/graph_edit.cpp:
```diff
@@ -531,65 +593,76 @@ void GraphEdit::_notification(int p_what) {
+	Rect2 frame_node_rect = p_node->get_rect();
 
-	Vector<GraphNode *> enclosed_nodes;
+	Vector<GraphControl *>
+			enclosed_nodes;
```
What happened here?

Geometror replied:
That's a good question :)

YuriSizov commented on scene/gui/graph_control.cpp:
```diff
@@ -0,0 +1,245 @@
+	}
+}
+
+#ifdef TOOLS_ENABLED
+void GraphControl::_edit_set_position(const Point2 &p_position) {
```
This should be at the top of the file, before other methods.

YuriSizov commented on scene/gui/graph_control.h:
```diff
@@ -0,0 +1,95 @@
+		return resizing;
+	}
+
+	GraphControl(){};
```
```suggestion
	GraphControl() {}
```

I think this is our convention for empty constructors.

YuriSizov reviewed (commented):
A few more notes from the main classes. So far I've ignored everything related to frames/comments, and I haven't looked through the updated GraphNode.

YuriSizov commented:
So, my overall impression so far is that we should split this into a few PRs (or at least a few commits, but with dedicated PRs we can merge them in steps, and not just review them easier).

- First, we clean up GraphEdit, refactoring the code and renaming the methods and properties. Since at this point we still have the old GraphNode code, methods that deal with frames/comments should be adjusted compared to the current state of this PR. General code improvements should still be made, but the relevant logic would remain more or less the same. Due to renames, some this commit would touch animation and shader files as well.
  - You can include snapping/grid and arranger improvements here as well.
- Second, we do a clean up on GraphNode, similar to what we did to GraphEdit. This can be done alongside the first step, if desired, but the key point is that GraphNode remains structurally the same, no base class, no frame/comment code extraction.
  - Here you would also remove the outdated code related to visual scripts.
- Last, we do the overhaul of nodes, split GraphNode into two, and remove everything related to frames/comments to be added with a future PR.

I believe, that this way we can review, merge, and iterate quicker, as right now this PR contains A LOT of changes which are codestyle/cosmetic. And for the most part those changes seem independent from logical changes, so factoring them out into dedicated PR doesn't appear challenging to me.

@Geometror Would you be up to follow with this plan? Does it sound sensible to you? If so, we could probably be done with this within a week. I'll be here to quickly review each step.

Geometror commented on scene/resources/default_theme/grid_toggle.svg:
Added it, but most of the icons in the default_theme folder don't have these attributes.

Geometror committed:
Renames and fixes addressing the review comments
(snap->snapping, idx->index/slot_index, toggle_grid->grid_toggle for theme icon, minor code style fixes)

Geometror committed:
Remaining compat-breaking/consmetic renames and minor fixes

Nyabinary commented:
What the milestone for this?

YuriSizov commented:
@Nyabinary 4.2. This was split into several PRs to make the process easier for everyone. You can see them linked above your comment.

MewPurPur commented:
<details>
<summary>Here are optimized versions for the new icons in this PR</summary>

GraphControl: `<svg height="16" width="16" viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg"><circle cx="8" cy="8" r="2" fill="#8eef97"/><circle cx="8" cy="8" r="5" stroke-width="2" stroke="#8eef97" fill="none"/></svg>`

GraphEdit: `<svg height="16" width="16" viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg"><path d="M11 1a1 1 0 0 0-1 1v3a1 1 0 0 0 1 1h3a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1zM6.732 5A2 2 0 0 1 7 6v1.117L9.268 6A2 2 0 0 1 9 5V3.883zM2 5a1 1 0 0 0-1 1v4a1 1 0 0 0 1 1h3a1 1 0 0 0 1-1V6a1 1 0 0 0-1-1zm5 3.883V10a2 2 0 0 1-.268 1L9 12.117V11a2 2 0 0 1 .268-1zM11 10a1 1 0 0 0-1 1v3a1 1 0 0 0 1 1h3a1 1 0 0 0 1-1v-3a1 1 0 0 0-1-1z" fill="#8eef97"/></svg>`

GraphFrame: `<svg height="16" width="16" viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg"><path d="M3 1a2 2 0 0 0-2 2v10a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3a2 2 0 0 0-2-2zm1.25 2h6.5A1.25 1.25 0 0 1 12 4.25V5a3 3 0 0 0 0 6v.75A1.25 1.25 0 0 1 10.75 13h-6.5A1.25 1.25 0 0 1 3 11.75v-7.5A1.25 1.25 0 0 1 4.25 3zM12 6a2 2 0 1 1 0 4 2 2 0 0 1 0-4z" fill="#8eef97"/></svg>`

GraphNode: `<svg height="16" width="16" viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg"><path d="M5 2a2 2 0 0 0-2 2h10a2 2 0 0 0-2-2zM3 5v1a3 3 0 0 1 0 6 2 2 0 0 0 2 2h6a2 2 0 0 0 2-2 3 3 0 0 1 0-6V5zm0 2a2 2 0 0 0 0 4 2 2 0 0 0 0-4zm10 0a2 2 0 0 0 0 4 2 2 0 0 0 0-4z" fill="#8eef97"/></svg>`

and GridToggle seems to already exist, so the PR adding it is probably a mistake.

YuriSizov commented:
This PR is superseded by https://github.com/godotengine/godot/pull/79311 and its companions. Everything outlined here is now implemented. Feel free to open follow-up PRs to fix bugs and add new features and enhancements!
